### PR TITLE
Create simulated jedec device to use in place of AP_Logger_SITL

### PR DESCRIPTION
**Pull request by shiv-tyagi**
This solves the issue #19128. We currently use AP_Logger_SITL to act like a block device. In this PR, we add a JEDEC dataflash simulator which use AP_Logger_Dataflash backend (the one we use on real hardware) to do this work. This eliminates the need of AP_Logger_SITL backend. 
Thanks @Divyateja04 for doing all the initial work on this one and @peterbarker for guiding us throughout.

### CODE REVIEW

File: libraries/AP_Logger/AP_Logger.h [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 // set default for HAL_LOGGING_DATAFLASH_ENABLED
 #ifndef HAL_LOGGING_DATAFLASH_ENABLED
-#define HAL_LOGGING_DATAFLASH_ENABLED 0
+#define HAL_LOGGING_DATAFLASH_ENABLED HAL_LOGGING_ENABLED
```
Comment (peterbarker) [inline libraries/AP_Logger/AP_Logger.h:14]:
```suggestion
#define HAL_LOGGING_DATAFLASH_ENABLED (CONFIG_HAL_BOARD == HAL_BOARD_SITL)
```

File: libraries/SITL/SIM_JEDEC.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                memset(fill, 0xFF, sizeof(fill));
+                const size_t write_ret = pwrite(storage_fd, fill, sizeof(fill), xfr_addr);
+                if (write_ret != sizeof(fill)) {
+                    printf("Failed sector erase");
+                }
+                write_enabled = false;
+                break;
+            }
+            default:
+                abort();
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:130]: panic instead, similarly elsehwere

File: libraries/SITL/SIM_JEDEC.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        switch (state) {
+        case State::WAITING: {
+            // find a command
+            uint8_t command = tx_buf[0];
+            switch (command) {
+            case JEDEC_RDID:
+                state = State::READING_RDID;
+                break;
+            case JEDEC_READ:
+                xfr_addr = tx_buf[1] << 16 | tx_buf[2] << 8 | tx_buf[3];
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:66]: We should probably add some sanity checks in here that we've actually got that many bytes in the transfer.

Similarly elsewhere

File: libraries/SITL/SIM_JEDEC.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                break;
+            }
+            case JEDEC_BULK_ERASE:  {
+                if (!write_enabled) {
+                    AP_HAL::panic("Writes not enabled");
+                }
+                if (xfr_addr + tfr.len > storage_size()) {
+                    abort();
+                }
+                uint8_t fill[storage_size()];
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:103]: Let's not put 32kB+ on the stack, please :-)

Loop instead, maybe a helper function for `SECTOR4_ERASE` and `BULK_ERASE`?

**Review by peterbarker — CHANGES_REQUESTED** [review]
Looks pretty reasonable, but this should work against the autotest's `DataFlashSITL` test.  We're not emitting a a key statustext which it requires ATM.

File: libraries/AP_Logger/AP_Logger.h (hunk 1 above)
Comment (shiv-tyagi) [reply libraries/AP_Logger/AP_Logger.h]: Done! Thanks

File: libraries/SITL/SIM_JEDEC.cpp (hunk 2 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp]: Done! Thanks

File: libraries/SITL/SIM_JEDEC.cpp (hunk 4 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp]: Done! Have split bulk erase into multiple block erases and block erases into multiple sector erases.

File: libraries/SITL/SIM_JEDEC.cpp (hunk 3 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp]: Done! Thanks @hendjoshsr71 for helping me do this.

**Comment by shiv-tyagi** [comment]
@peterbarker I have resolved all the concerns raised by you. We are now passing all the tests too. Please check and let me know if I need to do anything more to get this merged. Thanks!

File: libraries/SITL/SIM_JEDEC.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+void JEDEC::open_storage_fd()
+{
+    if (storage_fd != -1) {
+        AP_HAL::panic("Should not have been called");
+    }
+    
+    const char *filepath = filename();
+    uint32_t flags = O_RDWR|O_CLOEXEC;
+    storage_fd = open(filepath, flags, 0777);
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:20]:
```suggestion
    storage_fd = open(filepath, flags, 0660);
```

... similarly elsewhere, and apologies if this is just copy/pasted :-)

File: libraries/SITL/SIM_JEDEC.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        AP_HAL::panic("open(%s): %s", filepath, strerror(errno));
+    }
+    if (ftruncate(storage_fd, storage_size()) != 0) {
+        AP_HAL::panic("truncate(%s): %s", filepath, strerror(errno));
+    }
+}
+
+void JEDEC::sector4k_erase (uint32_t addr)
+{
+    uint32_t fill_length = get_page_size()*get_page_per_sector();
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:38]: How big can this get?

We don't want 4k on the stack; try to limit any fill buffer to, say, 512 bytes.

File: libraries/SITL/SIM_JEDEC.cpp [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+                write_enabled = true;
+                break;
+            case JEDEC_RDSR:
+                state = State::READING_RDSR;
+                break;
+            case JEDEC_SECTOR4_ERASE:   {
+                xfr_addr = parse_addr(tx_buf, tfr.len);
+                if (!write_enabled) {
+                    AP_HAL::panic("Writes not enabled");
+                }
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:122]:
```suggestion
                assert_writes_enabled();
```

... and use it elsewhere (search for "Writes not enabled"

File: libraries/SITL/SIM_JEDEC.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+            }
+            break;
+        }
+        case State::READING_RDID:
+            fill_rdid(rx_buf, tfr.len);
+            state = State::WAITING;
+            break;
+        case State::READING_RDSR:
+            fill_rdsr(rx_buf, tfr.len);
+            state=State::WAITING;
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:155]:
```suggestion
            state = State::WAITING;
```

File: libraries/AP_HAL_SITL/SITL_cmdline.cpp [hunk 9]
```diff
@@ -581,9 +580,6 @@ void SITL_State::_parse_command_line(int argc, char * const argv[])
 
     if (erase_all_storage) {
         AP_Param::erase_all();
-#if HAL_LOGGING_SITL_ENABLED
-        unlink(AP_Logger_SITL::filename);
```
Comment (peterbarker) [inline libraries/AP_HAL_SITL/SITL_cmdline.cpp:585]: We still need to do this; there's a method on halsitl which allows you to check if you need to clear storage.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_SITL/SITL_cmdline.cpp (hunk 9 above)
Comment (peterbarker) [reply libraries/AP_HAL_SITL/SITL_cmdline.cpp:585]: See https://github.com/ardupilot/ardupilot/blob/master/libraries/SITL/SIM_RAMTRON.cpp#L19

File: libraries/SITL/SIM_JEDEC.cpp (hunk 5 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp:20]: Done!

File: libraries/SITL/SIM_JEDEC.cpp (hunk 6 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp:38]: I have split sector erase into multiple page erases. Now it used a fill buffer of 256 bytes only.

File: libraries/SITL/SIM_JEDEC.cpp (hunk 7 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp:122]: Done!

File: libraries/SITL/SIM_JEDEC.cpp (hunk 8 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp:155]: Done!

File: libraries/AP_HAL_SITL/SITL_cmdline.cpp (hunk 9 above)
Comment (shiv-tyagi) [reply libraries/AP_HAL_SITL/SITL_cmdline.cpp:585]: Done!

**Review by peterbarker — CHANGES_REQUESTED** [review]
This needs to pass:
```
./Tools/autotest/autotest.py build.Rover test.Rover.DataFlashSITL --valgrind --debug
```

ATM it gives me a timeout - may be running out of space before finishing the test or something?

File: libraries/AP_HAL_SITL/SITL_cmdline.cpp (hunk 9 above)
Comment (shiv-tyagi) [reply libraries/AP_HAL_SITL/SITL_cmdline.cpp:585]: As discussed on discord, I have removed this from here and handled this in `init()` method of JEDEC.

File: libraries/SITL/SIM_JEDEC.cpp [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+{
+    if (storage_fd != -1) {
+        AP_HAL::panic("Should not have been called");
+    }
+    const char *filepath = filename();
+    uint32_t flags = O_RDWR|O_CREAT;
+    if (hal.get_wipe_storage()) {
+        unlink(filepath);
+    }
+    storage_fd = open(filepath, flags, 0644);
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:22]:
```suggestion
    if (hal.get_wipe_storage()) {
        unlink(filepath);
    }
    storage_fd = open(filepath, O_RDWR|O_CREAT, 0644);
```

File: libraries/SITL/SIM_JEDEC.cpp [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    }
+    if (ftruncate(storage_fd, storage_size()) != 0) {
+        AP_HAL::panic("truncate(%s): %s", filepath, strerror(errno));
+    }
+}
+
+void JEDEC::sector4k_erase (uint32_t addr)
+{
+    for (uint8_t i=0; i<get_page_per_sector(); i++) {
+        page_erase(addr + i*get_page_size());
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:34]: `page_erase` could take a page number, and `sector4k` could take a sector number - and check their alignment.

Same  with `block64k_erase`

Doesn't need to be done as part of this PR.

File: libraries/SITL/SIM_JEDEC.cpp [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+{
+    // we have 16 sectors in a block
+    for (uint16_t i=0; i<16; i++) {
+        sector4k_erase(addr + i*get_page_per_sector()*get_page_size());
+    }
+}
+
+void JEDEC::page_erase (uint32_t addr)
+{
+    uint32_t fill_length = get_page_size();
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC.cpp:48]:
```suggestion
    const uint32_t fill_length = get_page_size();
```

File: libraries/SITL/SIM_JEDEC_MX25L3206E.h [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+    void fill_rdid(uint8_t *buf, uint8_t len) override;
+    void fill_rdsr(uint8_t *buf, uint8_t len) override;
+
+    const char *filename() const override { return "dataflash.bin"; }
+
+    uint32_t storage_size() const override {
+        // page size * page per block * no. of blocks = 256 * 256 * 64 = 4194304
+        return 4194304;
+    }
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC_MX25L3206E.h:19]: Why not do this in the parent class, or as a `constexpr` based on the othe callbacks?

File: libraries/SITL/SIM_JEDEC_MX25L3206E.h [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+    uint8_t get_num_blocks() const override { return 64; }
+    uint16_t get_page_per_block() const override { return 256; }
+    uint8_t get_page_per_sector() const override { return 16; }
+    uint16_t get_page_size() const override { return 256; }
+    uint32_t get_num_pages() const override { return 16384; } // get_num _blocks()*get_page_per_block()
```
Comment (peterbarker) [inline libraries/SITL/SIM_JEDEC_MX25L3206E.h:25]: `constexpr` or parent class?

**Review by peterbarker — CHANGES_REQUESTED** [review]
We need to not change the default backend behaviour for real boards and remove the dead code - but the other comments in here don't really need addressing.

File: libraries/AP_Logger/AP_Logger.cpp [hunk 15]
```diff
@@ -45,10 +44,10 @@ extern const AP_HAL::HAL& hal;
 #endif
 
 #ifndef HAL_LOGGING_BACKENDS_DEFAULT
-# if HAL_LOGGING_DATAFLASH_ENABLED
-#  define HAL_LOGGING_BACKENDS_DEFAULT Backend_Type::BLOCK
-# elif HAL_LOGGING_FILESYSTEM_ENABLED
+# if HAL_LOGGING_FILESYSTEM_ENABLED
 #  define HAL_LOGGING_BACKENDS_DEFAULT Backend_Type::FILESYSTEM
+# elif HAL_LOGGING_DATAFLASH_ENABLED
+#  define HAL_LOGGING_BACKENDS_DEFAULT Backend_Type::BLOCK
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply libraries/AP_Logger/AP_Logger.cpp]: Fixed it!

File: libraries/SITL/SIM_JEDEC.cpp (hunk 10 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp]: Done!

File: libraries/SITL/SIM_JEDEC.cpp (hunk 12 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC.cpp]: Done!

File: libraries/SITL/SIM_JEDEC_MX25L3206E.h (hunk 13 above)
Comment (shiv-tyagi) [reply libraries/SITL/SIM_JEDEC_MX25L3206E.h]: Moved to parent class

File: libraries/AP_HAL_SITL/SITL_cmdline.cpp (hunk 9 above)
Comment (shiv-tyagi) [reply libraries/AP_HAL_SITL/SITL_cmdline.cpp:585]: removed it now

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**